portal.go

```diff
@@ -2047,22 +2125,65 @@ func (portal *Portal) HandleiMessage(msg *imessage.Message) id.EventID {
 		portal.bridge.IM.SendMessageBridgeResult(msg.ChatGUID, msg.GUID, eventID, overrideSuccess || hasMXID)
 	}()
 
+	// Look up the message in the database
+	dbMessage = portal.bridge.DB.Message.GetLastByGUID(portal.GUID, msg.GUID)
+
 	if portal.IsPrivateChat() && msg.ChatGUID != portal.LastSeenHandle {
 		portal.log.Debugfln("Updating last seen handle from %s to %s", portal.LastSeenHandle, msg.ChatGUID)
 		portal.LastSeenHandle = msg.ChatGUID
 		portal.Update(nil)
 	}
 
+	// Handle message tapbacks
 	if msg.Tapback != nil {
 		portal.HandleiMessageTapback(msg)
 		return ""
-	} else if portal.bridge.DB.Message.GetLastByGUID(portal.GUID, msg.GUID) != nil {
-		portal.log.Debugln("Ignoring duplicate message", msg.GUID)
-		// Send a success confirmation since it's a duplicate message
-		overrideSuccess = true
+	}
+
+	// If the message exists in the database, handle edits or retractions
+	if dbMessage != nil && dbMessage.MXID != "" {
+
+		// Clean up the text so the length is accurate
+		msg.Text = strings.ReplaceAll(msg.Text, "\ufffc", "")
+		msg.Subject = strings.ReplaceAll(msg.Subject, "\ufffc", "")
+
+		// DEVNOTE: It seems sometimes the message is just edited to remove data instead of actually retracting it
+		if msg.IsRetracted ||
+			(len(msg.Attachments) == 0 && len(msg.Text) == 0 && len(msg.Subject) == 0) {
+
+			// Retract existing message
+			if portal.HandleMessageRevoke(*msg) {
+				portal.zlog.Debug().Str("messageGUID", msg.GUID).Str("chatGUID", msg.ChatGUID).Msg("Revoked message")
+			} else {
+				portal.zlog.Warn().Str("messageGUID", msg.GUID).Str("chatGUID", msg.ChatGUID).Msg("Failed to revoke message")
+			}
+
+			overrideSuccess = true
+		} else if msg.IsEdited {
+
```

Comment:
I think the easiest way to prevent multipart messages from causing issues is to drop edits here if `dbMessage.Part > 0` (because that means there are multiple parts). Shouldn't need to change any other code